### PR TITLE
[GRPO]: Fix Multi-GPU training for Entropy based masking of tokens.

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?
#3933 identified that training a model with `top_entropy_quntile < 1` causes the training to hang indefinitely in a multi-gpu setting because the `gather` operation to gather the non_padded entropy tensor can have different shapes on different devices.

This PR fixes the said issue by:
1. Gathering the number of non padded entropy tokens.
2. Find the max number of non padded entropy tokens across all gpus.
3. Pad the non padded entropies to the max length
4. Gather all the entropies post padding.
5. Compute the threshold

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by pramodith** [comment]
Tested that training works correctly on a machine with 2 A40s. using this script:

<details>
<summary>Click to expand code</summary>

```
from trl import GRPOConfig, GRPOTrainer
from datasets import load_dataset
from accelerate import PartialState
from transformers import AutoModelForCausalLM, AutoTokenizer

dataset = load_dataset("trl-lib/tldr", split="train[:50]")

def reward_len(completions, **kwargs):
    return [-abs(20 - len(completion)) for completion in completions]

training_args = GRPOConfig(
    max_steps=10,
    per_device_train_batch_size=8,
    num_generations=8,
    logging_steps=1,
    gradient_checkpointing=True,
    gradient_checkpointing_kwargs = {"use_reentrant": False}, #must be false for DDP
    top_entropy_quantile=0.2,
    max_completion_length=1024
)

trainer = GRPOTrainer(
    model="Qwen/Qwen2-0.5B-Instruct",
    reward_funcs=reward_len,
    args=training_args,
    train_dataset=dataset,
)
trainer.train()
```

```
accelerate launch --num_processes 2 --multi_gpu --mixed_precision=bf16 examples/scripts/test.py
```

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1058,9 +1058,37 @@ def get_high_entropy_mask(
...
+                    torch.zeros(
+                        max_non_pad_entropies_seq_length - non_pad_entropies_seq_length.item(),
+                        device=non_pad_entropies.device,
+                    ),
+                ]
+            )
+            all_padded_entropies = self.accelerator.gather(padded_entropies)
+            all_padded_entropies_mask = self.accelerator.gather(padded_entropies_mask)
+            all_non_padded_entropies = all_padded_entropies[all_padded_entropies_mask.bool()].flatten()
+        else:
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: do we need this if/else?

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: I think it can be removed

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1058,9 +1058,37 @@ def get_high_entropy_mask(
...
+            # The shape of non_pad_entropies can be different on each gpu/device.
+            # this can cause the gather operation to hang. So we first gather the lengths
+            # of non_pad_entropies and pad them to the max length before doing a gather.
+            non_pad_entropies_seq_length = torch.tensor([non_pad_entropies.numel()], device=entropies.device)
+            max_non_pad_entropies_seq_length = self.accelerator.gather(non_pad_entropies_seq_length).max().item()
+            padded_entropies = torch.cat(
+                [
+                    non_pad_entropies,
+                    torch.zeros(
+                        max_non_pad_entropies_seq_length - non_pad_entropies_seq_length.item(),
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: probably simpler?
```suggestion
                        max_non_pad_entropies_seq_length - non_pad_entropies.numel()
```

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: Ummm there are a few ops that aren't useful to have in the single gpu-case right? Like all the first four lines in the if block aren't relevant/needed for a single gpu.

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: they'll probably be no-op, no?

**Comment by qgallouedec** [comment]
just merge main to your branch to fix the CI

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: These two won't be no-op 

```
 non_pad_entropies_seq_length = torch.tensor([non_pad_entropies.numel()], device=entropies.device)
 max_non_pad_entropies_seq_length = self.accelerator.gather(non_pad_entropies_seq_length).max().item()

torch.zeros(
                        max_non_pad_entropies_seq_length - non_pad_entropies.numel(),
                        device=non_pad_entropies.device,
                    ),
```

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: indeed but that's ok I think

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -762,9 +762,37 @@ def get_high_entropy_mask(
...
+                    torch.ones_like(non_pad_entropies),
+                    torch.zeros(
+                        max_non_pad_entropies_seq_length - non_pad_entropies_seq_length.item(),
+                        device=non_pad_entropies.device,
+                    ),
+                ]
+            )
+            all_padded_entropies = self.accelerator.gather(padded_entropies)
+            all_padded_entropies_mask = self.accelerator.gather(padded_entropies_mask)
+            all_non_padded_entropies = all_padded_entropies[all_padded_entropies_mask.bool()].flatten()
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
            all_non_padded_entropies = all_padded_entropies[all_padded_entropies_mask.bool()]
```

already flat

**Comment by qgallouedec** [comment]
can you also remove `accelerator=None` from the function signature?
